subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java

```diff
@@ -116,14 +123,22 @@ private Configuration createTestDataVariant(Project project, JvmTestSuite suite,
             attributes.attribute(DocsType.DOCS_TYPE_ATTRIBUTE, objects.named(DocsType.class, DocsType.TEST_RESULTS));
             attributes.attribute(Verification.TEST_SUITE_NAME_ATTRIBUTE, objects.named(Verification.class, suite.getName()));
             attributes.attribute(Verification.TARGET_NAME_ATTRIBUTE, objects.named(Verification.class, suite.getName()));
-            attributes.attributeProvider(TestType.TEST_TYPE_ATTRIBUTE, suite.getTestType().map(tt -> objects.named(TestType.class, tt)));
+            attributes.attributeProvider(TestType.TEST_TYPE_ATTRIBUTE, suite.getTestType().map(createNamedTestTypeAndVerifyUniqueness(project, suite)));
         });
 
         variant.getOutgoing().artifact(
             target.getTestTask().flatMap(AbstractTestTask::getBinaryResultsDirectory),
             artifact -> artifact.setType(ArtifactTypeDefinition.DIRECTORY_TYPE)
         );
+    }
 
-        return variant;
+    private Transformer<TestType, String> createNamedTestTypeAndVerifyUniqueness(Project project, TestSuite suite) {
+        return tt -> {
+            final TestSuite other = testTypesInUse.putIfAbsent(tt, suite);
```

Comment:
From a functional programming POV, I don't like that this lambda has a side-effect of mutating `testTypesInUse`. But I don't think there's another way to do this. Just curious: did you discuss any alternative approaches with Louis or Sterling?

Reply:
Not at the written code level, just at the conceptual level.  I agree it's not ideal, but it works, and it is all only internal details which could be changed later.  I'm open to suggestions how to improve it, but after 5 minutes of trying I can't think of a better way.